src/main/java/com/hillel/items_exchange/service/AdvertisementService.java

```diff
@@ -52,24 +51,64 @@ public AdvertisementDto createAdvertisement(AdvertisementDto advertisementDto, U
         Advertisement adv = mapDtoToAdvertisement(advertisementDto);
         adv.setUser(user);
         adv.setStatus(Status.NEW);
-        adv.getProduct().setSubcategory(subcategoryService.findById(advertisementDto.getProduct().getSubcategoryId())
-                .orElseThrow(EntityNotFoundException::new));
+        updateSubcategory(adv.getProduct(), advertisementDto.getProduct().getSubcategoryId());
+
         return mapAdvertisementToDto(advertisementRepository.save(adv));
     }
 
     public AdvertisementDto updateAdvertisement(AdvertisementDto dto) {
         Advertisement toUpdate = mapDtoToAdvertisement(dto);
         Advertisement fromDB = advertisementRepository.findById(dto.getId())
                 .orElseThrow(EntityNotFoundException::new);
-        BeanUtils.copyProperties(toUpdate, fromDB, "location", "user", "product", "created",
-                "updated", "status");
+
+        updateAdvertisement(toUpdate, fromDB);
+
+        Product toUpdateProduct = toUpdate.getProduct();
+        Product fromDBProduct = fromDB.getProduct();
+        updateProduct(toUpdateProduct, fromDBProduct);
+        updateSubcategory(fromDBProduct, toUpdateProduct.getSubcategory().getId());
+        updateImages(toUpdateProduct, fromDBProduct);
+        createOrUpdateLocation(toUpdate, fromDB);
+
         fromDB.setStatus(Status.UPDATED);
-        fromDB.getProduct().setSubcategory(subcategoryService.findById(dto.getProduct().getSubcategoryId())
-                .orElseThrow(EntityNotFoundException::new));
         Advertisement updatedAdvertisement = advertisementRepository.saveAndFlush(fromDB);
         return mapAdvertisementToDto(updatedAdvertisement);
     }
 
+    private void updateAdvertisement(Advertisement toUpdate, Advertisement fromDB) {
+        if (!fromDB.equals(toUpdate)) {
+            BeanUtils.copyProperties(toUpdate, fromDB, "created", "updated", "status", "location", "user", "product");
+        }
+    }
+
+    private void updateProduct(Product toUpdateProduct, Product fromDBProduct) {
+        if (!fromDBProduct.equals(toUpdateProduct)) {
+            BeanUtils.copyProperties(toUpdateProduct, fromDBProduct, "advertisement", "subcategory", "images");
+        }
+    }
+
+    private void updateSubcategory(Product fromDBProduct, long id) {
+        fromDBProduct.setSubcategory(subcategoryService.findById(id)
+                .orElseThrow(EntityNotFoundException::new));
+    }
+
+    private void updateImages(Product toUpdateProduct, Product fromDBProduct) {
+        List<Image> existImages = fromDBProduct.getImages();
+        List<Image> newImages = toUpdateProduct.getImages();
+        if (existImages.size() != newImages.size() || !existImages.containsAll(newImages)) {
+            existImages.retainAll(newImages);
+        }
+    }
```

Comment:
In fact this `updateImages` method doesn't update images, it only removes elements from the `existImages` that are not contained in the `newImages`. Changing field value defaultPhoto will be prosseced within another task?

Reply:
Yup, I missed that part, thanks. I'll update it.